detekt-rules-coroutines/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/coroutines/CoroutineLaunchedInTestWithoutRunTest.kt

```diff
@@ -0,0 +1,147 @@
+package io.gitlab.arturbosch.detekt.rules.coroutines
+
+import io.gitlab.arturbosch.detekt.api.CodeSmell
+import io.gitlab.arturbosch.detekt.api.Config
+import io.gitlab.arturbosch.detekt.api.Entity
+import io.gitlab.arturbosch.detekt.api.Issue
+import io.gitlab.arturbosch.detekt.api.RequiresTypeResolution
+import io.gitlab.arturbosch.detekt.api.Rule
+import io.gitlab.arturbosch.detekt.rules.fqNameOrNull
+import io.gitlab.arturbosch.detekt.rules.hasAnnotation
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
+import org.jetbrains.kotlin.psi.KtExpression
+import org.jetbrains.kotlin.psi.KtNamedFunction
+import org.jetbrains.kotlin.psi.psiUtil.anyDescendantOfType
+import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.calls.util.getCalleeExpressionIfAny
+import org.jetbrains.kotlin.resolve.calls.util.getResolvedCall
+import org.jetbrains.kotlin.resolve.calls.util.getType
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.resolve.source.getPsi
+import org.jetbrains.kotlin.utils.addToStdlib.ifTrue
+
+/**
+ * Detect coroutine launches from `@Test` functions outside a `runTest` block.
+ * Launching coroutines in tests without using `runTest` could potentially swallow exceptions,
+ * altering test results or causing crashes on other unrelated tests.
+ *
+ * <noncompliant>
+ * @@Test
+ * fun `test that launches a coroutine`() {
+ *     val scope = CoroutineScope(Dispatchers.Unconfined)
+ *     scope.launch {
+ *         suspendFunction()
+ *     }
+ * }
+ * </noncompliant>
+ *
+ * <compliant>
+ * @@Test
+ * fun `test that launches a coroutine`() = runTest {
+ *     launch {
+ *         suspendFunction()
+ *     }
+ * }
+ * </compliant>
+ *
+ */
+@RequiresTypeResolution
+class CoroutineLaunchedInTestWithoutRunTest(config: Config) : Rule(config) {
+    private val funCoroutineLaunchesTraverseHelper = FunCoroutineLaunchesTraverseHelper()
+
+    override val issue = Issue(
+        javaClass.simpleName,
+        description = "Launching coroutines in tests without a `runTest` block could swallow exceptions. " +
+            "You should use `runTest` to avoid altering test results."
+    )
+
+    override fun visitNamedFunction(initialFunction: KtNamedFunction) {
+        if (!initialFunction.hasBody()) return
+        if (!initialFunction.hasAnnotation(TEST_ANNOTATION_NAME)) return
+        if (initialFunction.runsInRunTestBlock()) return
+
+        funCoroutineLaunchesTraverseHelper
+            .isFunctionOrChildrenLaunchingCoroutines(initialFunction, bindingContext)
+            .ifTrue { report(CodeSmell(issue, Entity.from(initialFunction), MESSAGE)) }
```

Comment:
```suggestion
        // By this point we know we're inside a test function that is not a `runTest` function.
        if (funCoroutineLaunchesTraverseHelper.isFunctionLaunchingCoroutines(initialFunction, bindingContext)) {
            report(CodeSmell(issue, Entity.from(initialFunction), MESSAGE))
        }
```

This improves the readability here